File: Online-judge/leetcode/two-sum-bsts.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

"""
Md.Shohag <shohag.fks@gmail.com>
problem: https://leetcode.com/problems/two-sum-bsts/description/
Time complexity O(m.logn)
Space complexity O(logm+longn)
"""
        
class Solution:
    def twoSumBSTs(self, root1: Optional[TreeNode], root2: Optional[TreeNode], target: int) -> bool:
        def binarySearch(node2, target2):
            if not node2:
                return False
            if node2.val == target2:
                return True
            elif node2.val > target2:
                return binarySearch(node2.left, target2)
            else:
                return binarySearch(node2.right, target2)
        
        def dfs(root, target):
            if not root:
                return False
            if binarySearch(root2, target - root.val):
                return True
            return dfs(root.left, target) or dfs(root.right, target)

        return dfs(root1, target)

File: Online-judge/leetcode/two-sum-bsts.cpp
# include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

/**
 * @brief 
 * Problem: https://leetcode.com/problems/two-sum-bsts/description/
 * Time complexity: O(m * n)
 * Space complexity: O(n + n)
 */

class Solution {
private:
    void dfs1(TreeNode* node, vector<int>& list) {
        if (!node) return;
        list.push_back(node->val);
        dfs1(node->left, list);
        dfs1(node->right, list);
    }

    void dfs2(TreeNode* node, unordered_set<int>& list) {
        if (!node) return;
        list.insert(node->val);
        dfs2(node->left, list);
        dfs2(node->right, list);
    }
    
public:
    bool twoSumBSTs(TreeNode* root1, TreeNode* root2, int target) {
        vector<int> list1;
        unordered_set<int> list2;

        dfs1(root1, list1);
        dfs2(root2, list2);

        for (int val : list1) {
            if (list2.find(target - val) != list2.end()) return true;
        }
        return false;
    }
};